# 영화별 leading role 배우 테이블 만들기
** leading role 정의: 한 영화의 출연배우를 보여주는 행들 중에 가장 윗행에 있는 배우

In [1]:
import pandas as pd

In [3]:
actors = pd.read_csv('../01.Data_Collection/actors.csv') 
actorcode = pd.read_csv('../01.Data_Collection/actor_code.csv')
actordetail = pd.read_csv('../01.Data_Collection/actor_detail.csv') 

In [4]:
actors.dtypes

cast         object
movieCd       int64
movieNm      object
peopleNm     object
repRoleNm    object
dtype: object

In [5]:
actorcode.dtypes

filmoNames    object
movieNm       object
peopleCd       int64
peopleNm      object
repRoleNm     object
dtype: object

In [6]:
actordetail.dtypes

movieCd        object
movieNm        object
moviePartNm    object
peopleCd        int64
peopleNm       object
sex            object
dtype: object

In [7]:
# 데이터타입 전부 string으로 맞춰주기
actors['movieCd'] = actors['movieCd'].apply(lambda x: str(x))
actorcode['peopleCd'] = actorcode['peopleCd'].apply(lambda x: str(x))
actordetail['peopleCd'] = actordetail['peopleCd'].apply(lambda x: str(x))

In [8]:
actors.dtypes

cast         object
movieCd      object
movieNm      object
peopleNm     object
repRoleNm    object
dtype: object

In [9]:
actors.isnull().sum()

cast         2062
movieCd         0
movieNm         0
peopleNm        0
repRoleNm       0
dtype: int64

In [10]:
actorcode.isnull().sum()

filmoNames    0
movieNm       0
peopleCd      0
peopleNm      0
repRoleNm     0
dtype: int64

In [11]:
actordetail.isnull().sum()

movieCd          0
movieNm          0
moviePartNm      0
peopleCd         0
peopleNm         0
sex            668
dtype: int64

### 'actors.csv' 파일의 'cast' 컬럼(배역명) 결측치 2062개  
일단은 이 컬럼을 사용하지 않을 것이므로 그냥 둔다.

### 'actordetail.csv' 파일의 'sex' 컬럼(성별) 결측치 668개  
각 영화 leading role(제1주연배우)의 성별 결측치만 채운다.  

In [12]:
# movieCd, movieNm으로 groupby 되어있는 듯한 형태로 영화에 나오는 모든 배우가 1인당 1행씩
actors.head()

cast   movieCd movieNm peopleNm repRoleNm
0   서도철  20148048     베테랑      황정민        배우
1   조태오  20148048     베테랑      유아인        배우
2  최 상무  20148048     베테랑      유해진        배우
3  오 팀장  20148048     베테랑      오달수        배우
4   미스봉  20148048     베테랑      장윤주        배우

In [13]:
# actors와 비슷한 형태인데 추가로 배우 코드와 배우의 필모그래피 확인 가능
actorcode.head()

filmoNames movieNm  peopleCd  \
0  인질(가제)|다만 악에서 구하소서|공작|군함도|군함도 감독판|아수라|곡성|검사외전|...     베테랑  10090290   
1  소리도 없이|국가부도의 날|버닝|좋아해줘|사도|베테랑|우아한 거짓말|우리별 일호와 ...     베테랑  10053802   
2  봉오동 전투|승리호(가제)|말모이|완벽한 타인|레슬러|1987|택시운전사|공조|럭키...     베테랑  10054128   
3  이웃 사촌|조선명탐정: 흡혈괴마의 비밀|신과함께-죄와 벌|니 부모 얼굴이 보고 싶다...     베테랑  10050151   
4                                                베테랑     베테랑  20211726   

  peopleNm repRoleNm  
0      황정민        배우  
1      유아인        배우  
2      유해진        배우  
3      오달수        배우  
4      장윤주        배우

In [14]:
# peopleCd, peopleNm으로 groupby 되어있는 듯한 형태로 배우의 필모그래피가 1작품당 1행씩
actordetail.head()

movieCd      movieNm moviePartNm  peopleCd peopleNm sex
0  20197922  다만 악에서 구하소서          배우  10090290      황정민  남자
1  20192194       인질(가제)          배우  10090290      황정민  남자
2  20179101           공작          배우  10090290      황정민  남자
3  20161481          군함도          배우  10090290      황정민  남자
4  20177063      군함도 감독판          배우  10090290      황정민  남자

In [15]:
len(actors)

11054

In [16]:
len(actorcode)

10421

In [17]:
actors['movieNm'].nunique()

770

In [18]:
actorcode['movieNm'].nunique()

684

actors에는 있는데 actorcode에는 없는 영화 86개

In [19]:
set(actors['movieNm'].unique()) - set(actorcode['movieNm'].unique())

{'개인교수: 심화학습 (무삭제)',
 '경계',
 '고백할 수 없는',
 '그랜드파더',
 '그빨로',
 '금지된 섹스, 달콤한 복수',
 '기억의 소리',
 '깡치',
 '나는 증인이다',
 '나의 살던 고향은',
 '나홀로 휴가',
 '눈이라도 내렸으면',
 '달에 부는 바람',
 '당신자신과 당신의 것',
 '대결',
 '독도의 영웅들',
 '돈 맛',
 '동거의 목적(무삭제판)',
 '두 개의 연애',
 '두 번째 스물',
 '딜쿠샤',
 '뜨거운 이웃 무삭제 감독판',
 '러브레따',
 '레나',
 '맛있는 만남',
 '멜리스',
 '무수단',
 '물숨',
 '미스터 쿠퍼',
 '배짱이들',
 '부릉! 부릉! 브루미즈: 스피더의 모험 일기',
 '브링 홈: 아버지의 땅',
 '블랙가스펠2: 소울을 찾는 여정가운데 만난 사람들',
 '블랙스톤',
 '비치온더비치',
 '사돈의 팔촌',
 '삼례',
 '설행_눈길을 걷다',
 '섬. 사라진 사람들',
 '소꿉놀이',
 '수색역',
 '스틸 플라워',
 '시발, 놈: 인류의 시작',
 '시소',
 '야근 대신 뜨개질',
 '야한 누나들',
 '야한밤 심야극장',
 '양치기들',
 '어떻게 헤어질까',
 '업사이드 다운',
 '엘 꼰도르 빠사',
 '연애담',
 '연애의 발동 : 상해 여자, 부산 남자',
 '우리 손자 베스트',
 '우리 연애의 이력',
 '우리에게 떡볶이를 먹을 권리가 있다',
 '우주의 크리스마스',
 '울보',
 '워킹 스트리트',
 '위켄즈',
 '인천상륙작전 : 익스텐디드 에디션(확장판)',
 '일어서는 인간',
 '작은 형',
 '제자도 - 제자,옥한흠2',
 '짐작보다 따뜻하게',
 '철원기행',
 '청춘학당2 기생난입야사',
 '초인',
 '춘몽',
 '치킨게임',
 '카이: 거울 호수의 전설',
 '커터',
 '커튼콜',
 '탐욕의 별',
 '파리의 한국남자',
 '폭력의 법칙: 나쁜 피 두 번째 이야기',
 '프랑스 영화처럼',
 '하숙집 2 (무삭제)',
 '하프',
 

In [20]:
actors[actors['movieNm']=='연애담']

cast   movieCd movieNm peopleNm repRoleNm
3701       윤주  20159286     연애담      이상희        배우
3702       지수  20159286     연애담      류아벨        배우
3703       영호  20159286     연애담      박근록        배우
3704       영은  20159286     연애담      임성미        배우
3705     지수아빠  20159286     연애담      김종수        배우
3706   Bar 손님  20159286     연애담      유우람        배우
3707  데이트술집손님  20159286     연애담      이성용        배우
3708       학생  20159286     연애담      이성용        배우
3709       교수  20159286     연애담      이지수        배우
3710      모텔남  20159286     연애담      장병기        배우
3711      주방장  20159286     연애담      차승호        배우

In [21]:
# 86개 영화의 출연자 필모그래피에 해당 영화가 있다 하더라도 따로 데이터행으로 안 들어가 있는 상태
actorcode[actorcode['peopleNm']=='류아벨']

filmoNames movieNm  peopleCd  \
35    샘|연애담|간신|베테랑|개를 훔치는 완벽한 방법|네버다이 버터플라이|초능력자|푸른 ...     베테랑  20236647   
654   샘|연애담|간신|베테랑|개를 훔치는 완벽한 방법|네버다이 버터플라이|초능력자|푸른 ...      간신  20236647   
9059  샘|연애담|간신|베테랑|개를 훔치는 완벽한 방법|네버다이 버터플라이|초능력자|푸른 ...       샘  20236647   

     peopleNm repRoleNm  
35        류아벨        배우  
654       류아벨        배우  
9059      류아벨        배우

### leading role 찾기

In [22]:
# 각 영화코드를 딱 하나씩만 리스트에 넣는다.
moviecode = list(actors['movieCd'].unique())
 
# moviecode에서 원소를 하나씩 빼고나면 actors 행과의 비교가 끝나기 전에 리스트가 비어서 'list index out of range' 에러가 나므로 마지막 원소로 0을 추가해 놓는다.
moviecode.append(0) 

leading_role = actors.copy()

for i in range(len(leading_role)):
    
    # 동일한 영화코드를 가진 행들 중 첫번째 데이터를 보존하고, moviecode 리스트에서 그 코드를 제거한다.
    if leading_role.loc[i, 'movieCd'] == moviecode[0]:
        moviecode.pop(0)
    
    # 두번째 데이터부터 데이터프레임에서 삭제한다.
    else:
        leading_role = leading_role.drop(pd.DataFrame(leading_role.loc[i]), axis=0)

In [23]:
leading_role

cast   movieCd           movieNm peopleNm repRoleNm
0             서도철  20148048               베테랑      황정민        배우
44            안옥윤  20148851                암살      전지현        배우
81            안상구  20149314              내부자들      이병헌        배우
118           NaN  20145503            검은 사제들      김윤석        배우
138           NaN  20148846                사도      송강호        배우
169           NaN  20134798              연평해전      김무열        배우
189           NaN  20148845                스물      김우빈        배우
211           NaN  20136068              히말라야      황정민        배우
232           NaN  20149570  조선명탐정 : 사라진 놉의 딸      김명민        배우
264           공길용  20148855              극비수사      김윤석        배우
295           NaN  20149569            악의 연대기      손현주        배우
316           NaN  20156557        탐정 : 더 비기닝      권상우        배우
350           NaN  20157464               더 폰      손현주        배우
380            종대  20136888           강남 1970      이민호        배우
436         구청여직원  20140703              장수상회     박해빛나        배우
473           홍이수  20143572           뷰티 인사이드      한효주        배우
531           NaN  20148188               쎄시봉      김윤석        배우
563           NaN  20142407            오늘의 연애      이승기        배우
578            엄마  20149266             차이나타운      김혜수        배우
610           NaN  20142402                간신      주지훈        배우
663           NaN  20145171                대호      최민식        배우
685            장우  20157465              그놈이다       주원        배우
702    우수제약 임원들 2  20142855            성난 변호사      오상철        배우
732           NaN  20148461              살인의뢰      김상경        배우
753           허무혁  20145504         특종: 량첸살인기      조정석        배우
768           하재관  20156748        열정같은소리하고있네      정재영        배우
788           NaN  20143513            미쓰 와이프      엄정화        배우
821           허삼관  20131262               허삼관      하정우        배우
848           NaN  20148841               헬머니      김수미        배우
892           NaN  20130566               무뢰한      김남길        배우
...           ...       ...               ...      ...       ...
10686         NaN  20197846              악질경찰      이선균        배우
10699         이찬우  20194456             양자물리학      박해수        배우
10712          영주  20180825               미성년      염정아        배우
10724          덕호  20185985        광대들: 풍문조작단      조진웅        배우
10738         장세출  20198453   롱 리브 더 킹: 목포 영웅      김래원        배우
10750        세종대왕  20182490             나랏말싸미      송강호        배우
10767         철수   20187322      힘을 내요, 미스터 리      차승원        배우
10778         유관순  20196601        항거:유관순 이야기      고아성        배우
10784         NaN  20180868                뺑반      공효진        배우
10802      이명준 대위  20193601     장사리 : 잊혀진 영웅들      김명민        배우
10813          용후  20180442                사자      박서준        배우
10837          장수  20184483              퍼펙트맨      설경구        배우
10844          세하  20185784         나의 특별한 형제      신하균        배우
10852          미수  20181416          유열의 음악앨범      김고은        배우
10862        호텔직원  20176251            내안의 그놈      오윤수        배우
10881         NaN  20184082               걸캅스      라미란        배우
10899      중수(삼촌)  20184543                변신      배성우        배우
10906         도일출  20184572       타짜: 원 아이드 잭      박정민        배우
10919          정일  20183096                생일      설경구        배우
10930         NaN  20177539               사바하      이정재        배우
10940         NaN  20184105               말모이      유해진        배우
10954          순호  20185881                증인      정우성        배우
10969          재훈  20195920         가장 보통의 연애      김래원        배우
10979          동수  20188941               악인전      마동석        배우
10986         황해철  20184462            봉오동 전투      유해진        배우
11000         박웅철  20188750      나쁜 녀석들: 더 무비      마동석        배우
11009         조일현  20176103                 돈      류준열        배우
11018          기택  20183782               기생충      송강호        배우
11030         NaN  20184621               엑시트      조정석        배우
11041         고반장  20182530              극한직업   

In [24]:
len(leading_role)

770

In [25]:
# 제목이 똑같은 영화는 없다.
leading_role['movieNm'].nunique()

770

In [26]:
leading_role['peopleNm'].nunique()

565

In [27]:
leading_role['repRoleNm'].unique()

array(['배우'], dtype=object)

In [28]:
actorcode.head(3)

filmoNames movieNm  peopleCd  \
0  인질(가제)|다만 악에서 구하소서|공작|군함도|군함도 감독판|아수라|곡성|검사외전|...     베테랑  10090290   
1  소리도 없이|국가부도의 날|버닝|좋아해줘|사도|베테랑|우아한 거짓말|우리별 일호와 ...     베테랑  10053802   
2  봉오동 전투|승리호(가제)|말모이|완벽한 타인|레슬러|1987|택시운전사|공조|럭키...     베테랑  10054128   

  peopleNm repRoleNm  
0      황정민        배우  
1      유아인        배우  
2      유해진        배우

In [29]:
leading_role = leading_role.merge(actorcode[['movieNm', 'peopleCd', 'peopleNm']], how='left', on=['movieNm', 'peopleNm'])
len(leading_role)

784

In [30]:
leading_role['movieNm'].value_counts()

소시민                             2
안시성                             2
너와 극장에서                         2
공동정범                            2
우리는 썰매를 탄다                      2
귀향, 끝나지 않은 이야기                  2
극장판 헬로카봇 : 백악기 시대               2
남자와 여자                          2
어른이 되면                          2
어떤이의 꿈                          2
무현, 두 도시 이야기 : 파이널 컷            2
콜리션                             2
뚜르: 내 생애 최고의 49일                2
망원동 인공위성                        2
최강전사 미니특공대 : 새로운 악당의 습격         1
수색역                             1
타투                              1
우주의 크리스마스                       1
생일                              1
굿바이 썸머                          1
런닝맨 : 풀룰루의 역습                   1
위대한 소원                          1
여자들                             1
오빠생각                            1
슈퍼 프렌즈                          1
리스펙트                            1
아기와 나                           1
나는 증인이다                         1
나만 없어 고양이                       1
우리 연애의 이력     

In [31]:
leading_role[leading_role['movieNm'] == '극장판 헬로카봇 : 백악기 시대'] 

cast   movieCd            movieNm peopleNm repRoleNm  peopleCd
571  차탄 목소리  20186501  극장판 헬로카봇 : 백악기 시대      이지현        배우  20237906
572  차탄 목소리  20186501  극장판 헬로카봇 : 백악기 시대      이지현        배우  20237906

In [32]:
# merge를 하면서 똑같은 데이터가 두 번 들어간 것도 있으니 중복값을 제거한다.
leading_role = leading_role.drop_duplicates()

In [33]:
len(leading_role)

770

In [34]:
leading_role.sample(3)

cast   movieCd      movieNm peopleNm repRoleNm  peopleCd
259  NaN  20154803    우리 연애의 이력      전혜빈        배우       NaN
405  NaN  20170163  밤섬해적단 서울불바다      권용만        배우  20203423
3    NaN  20145503       검은 사제들      김윤석        배우  10005276

In [35]:
leading_role.isnull().sum()

cast         423
movieCd        0
movieNm        0
peopleNm       0
repRoleNm      0
peopleCd      86
dtype: int64

### leading_role의 peopleCd 컬럼 (영화인코드) 86개의 결측치 채우기

In [37]:
peoplecd_nan = leading_role[leading_role['peopleCd'].isnull()]
peoplecd_nan

cast   movieCd                   movieNm peopleNm repRoleNm  \
229           NaN  20162022                        대결      이주승        배우   
230           NaN  20166127                     그랜드파더      박근형        배우   
231           장학수  20160041   인천상륙작전 : 익스텐디드 에디션(확장판)      이정재        배우   
232           NaN  20163098             카이: 거울 호수의 전설      김영은        배우   
233            윤주  20159286                       연애담      이상희        배우   
234           NaN  20161722                        히야      안보현        배우   
235           이혜리  20144441                섬. 사라진 사람들      박효주        배우   
236            세준  20152746                        커터      최태준        배우   
237           NaN  20169730               당신자신과 당신의 것      김주혁        배우   
238       신유화 중위   20152541                       무수단      이지아        배우   
239           NaN  20133071              브링 홈: 아버지의 땅    텐진 릭돌        배우   
240           NaN  20162541                        춘몽      한예리        배우   
241           최가인  20157441                       멜리스      홍수아        배우   
242           강민구  20155301                   두 번째 스물      김승우        배우   
243           NaN  20162608                 나의 살던 고향은      김용옥        배우   
244           NaN  20164527  부릉! 부릉! 브루미즈: 스피더의 모험 일기      양정화        배우   
245           NaN  20161694             제자도 - 제자,옥한흠2      권오중        배우   
246           NaN  20142706                        시소      이동우        배우   
247           NaN  20152369                    나홀로 휴가      박혁권        배우   
248          나레이션  20162969                        물숨      고희영        배우   
249            수현  20155388                        초인      채서진        배우   
250           NaN  20155390                      양치기들      박종환        배우   
251           마리아  20158822                 설행_눈길을 걷다      박소담        배우   
252           NaN  20157064                       커튼콜      장현성        배우   
253            엄마  20158824                  프랑스 영화처럼      이영란        배우   
254           NaN  20158359                  할머니의 먼 집      박삼순        배우   
255  故 박성빈 양 아버지   20152402                   업사이드 다운      박영우        배우   
256           NaN  20141589                      철원기행      문창길        배우   
257           NaN  20159447                        레나      박기림        배우   
258            승우  20158821                        삼례      이선호        배우   
..            ...       ...                       ...      ...       ...   
285           정은경  20158744                 짐작보다 따뜻하게       이선        배우   
286            태익  20159627                    사돈의 팔촌      장인섭        배우   
287           NaN  20140723                  달에 부는 바람      김예지        배우   
288            이섭  20158742                        울보      장유상        배우   
289            수민  20155387                        혼자      이주원        배우   
290           NaN  20158743                 눈이라도 내렸으면      강성국        배우   
291           NaN  20165164                      러브레따      예수정        배우   
292          내레이션  20162684                     탐욕의 별      김의성        배우   
293            태성  20162546                   워킹 스트리트      백성현        배우   
294           NaN  20167102                        휴가       고원        배우   
295           NaN  20162970     폭력의 법칙: 나쁜 피 두 번째 이야기      김영무        배우   
296           NaN  20166912          개인교수: 심화학습 (무삭제)      허예창        배우   
297            승구  20166652              동거의 목적(무삭제판)      제동화        배우   
298           NaN  20162145              청춘학당2 기생난입야사      전현수        배우   
299            민아  20151224                        하프      안용준        배우   
300           NaN  20126669                  엘 꼰도르 빠사      강예원        배우   
301           NaN  20150982            금지된 섹스, 달콤한 복수      송은채        배우   
302           NaN  20161584                    야한 누나들      이채담        배우   
303           NaN  20162005                    맛있는 만남      차주현        배우   
304           NaN  20162004                  야한밤 심야극장      이채담        배우   
305           NaN  20166867            뜨거운 이웃 무삭제 감독판      이채

In [38]:
actordetail.head(3)

movieCd      movieNm moviePartNm  peopleCd peopleNm sex
0  20197922  다만 악에서 구하소서          배우  10090290      황정민  남자
1  20192194       인질(가제)          배우  10090290      황정민  남자
2  20179101           공작          배우  10090290      황정민  남자

In [39]:
peoplecd_extrait = actordetail[(actordetail['movieCd'].isin(peoplecd_nan['movieCd'])) & (actordetail['peopleNm'].isin(peoplecd_nan['peopleNm']))]

In [40]:
peoplecd_extrait = peoplecd_extrait.drop_duplicates()

In [41]:
len(peoplecd_extrait)

64

In [42]:
peoplecd_extrait.head()

movieCd                  movieNm moviePartNm  peopleCd peopleNm sex
349   20158822                설행_눈길을 걷다          배우  20209686      박소담  여자
393   20144441               섬. 사라진 사람들          배우  20136191      배성우  남자
401   20158745                 고백할 수 없는          배우  20136191      배성우  남자
1287  20160041  인천상륙작전 : 익스텐디드 에디션(확장판)          배우  10057315      이정재  남자
1601  20169730              당신자신과 당신의 것          배우  10005383      김의성  남자

actordetail에서 86개 중 64개(중복값 포함) 가져옴

In [43]:
peoplecd_extrait.isnull().sum()

movieCd        0
movieNm        0
moviePartNm    0
peopleCd       0
peopleNm       0
sex            0
dtype: int64

In [44]:
leading_role.head()

cast   movieCd movieNm peopleNm repRoleNm  peopleCd
0  서도철  20148048     베테랑      황정민        배우  10090290
1  안옥윤  20148851      암살      전지현        배우  10061467
2  안상구  20149314    내부자들      이병헌        배우  10055626
3  NaN  20145503  검은 사제들      김윤석        배우  10005276
4  NaN  20148846      사도      송강호        배우  10037018

In [45]:
leading_role = leading_role.merge(peoplecd_extrait[['movieCd', 'peopleNm', 'peopleCd']], how='left', on=['movieCd', 'peopleNm'])
leading_role.head()

cast   movieCd movieNm peopleNm repRoleNm peopleCd_x peopleCd_y
0  서도철  20148048     베테랑      황정민        배우   10090290        NaN
1  안옥윤  20148851      암살      전지현        배우   10061467        NaN
2  안상구  20149314    내부자들      이병헌        배우   10055626        NaN
3  NaN  20145503  검은 사제들      김윤석        배우   10005276        NaN
4  NaN  20148846      사도      송강호        배우   10037018        NaN

In [46]:
len(leading_role)

774

In [47]:
leading_role = leading_role.drop_duplicates()
len(leading_role)

770

In [48]:
leading_role['peopleCd_x'] = leading_role['peopleCd_x'].fillna(leading_role['peopleCd_y'])

중복값 제외하고 결측치 38개 남음

In [49]:
leading_role.isnull().sum()

cast          423
movieCd         0
movieNm         0
peopleNm        0
repRoleNm       0
peopleCd_x     38
peopleCd_y    722
dtype: int64

In [50]:
leading_role = leading_role.drop('peopleCd_y', axis=1)
leading_role = leading_role.rename(columns={'peopleCd_x':'peopleCd'})

In [51]:
leading_role.head()

cast   movieCd movieNm peopleNm repRoleNm  peopleCd
0  서도철  20148048     베테랑      황정민        배우  10090290
1  안옥윤  20148851      암살      전지현        배우  10061467
2  안상구  20149314    내부자들      이병헌        배우  10055626
3  NaN  20145503  검은 사제들      김윤석        배우  10005276
4  NaN  20148846      사도      송강호        배우  10037018

38개는 성별 컬럼 붙이고 나서 해결하자.

In [52]:
leading_role = leading_role.merge(actordetail[['movieCd', 'peopleCd', 'sex']], how='left', on=['movieCd', 'peopleCd']).drop_duplicates()
leading_role.head()

cast   movieCd movieNm peopleNm repRoleNm  peopleCd sex
0   서도철  20148048     베테랑      황정민        배우  10090290  남자
7   안옥윤  20148851      암살      전지현        배우  10061467  여자
8   안상구  20149314    내부자들      이병헌        배우  10055626  남자
16  NaN  20145503  검은 사제들      김윤석        배우  10005276  남자
23  NaN  20148846      사도      송강호        배우  10037018  남자

In [53]:
len(leading_role)

770

In [54]:
leading_role.isnull().sum()

cast         423
movieCd        0
movieNm        0
peopleNm       0
repRoleNm      0
peopleCd      38
sex           56
dtype: int64

### 'cast' 컬럼(배역명) 결측치 423개
일단은 이 컬럼을 사용하지 않을 것이므로 그냥 둔다.

### 'peopleCd' 컬럼(영화인코드) 결측치 38개와 'sex' 컬럼(성별) 결측치 56개
KOBIS 홈페이지에서 검색해서 값을 넣는다.

In [55]:
leading_role[(leading_role['peopleCd'].isnull()) | (leading_role['sex'].isnull())]

cast   movieCd                movieNm peopleNm repRoleNm  \
506             출연  20143726               명령불복종 교사      최혜원        배우   
972             세준  20152746                     커터      최태준        배우   
983        신유화 중위   20152541                    무수단      이지아        배우   
984            NaN  20133071           브링 홈: 아버지의 땅    텐진 릭돌        배우   
991            최가인  20157441                    멜리스      홍수아        배우   
993            NaN  20162608              나의 살던 고향은      김용옥        배우   
998            NaN  20161694          제자도 - 제자,옥한흠2      권오중        배우   
999            NaN  20142706                     시소      이동우        배우   
1007          나레이션  20162969                     물숨      고희영        배우   
1026            엄마  20158824               프랑스 영화처럼      이영란        배우   
1027           NaN  20158359               할머니의 먼 집      박삼순        배우   
1028  故 박성빈 양 아버지   20152402                업사이드 다운      박영우        배우   
1029           NaN  20141589                   철원기행      문창길        배우   
1030           NaN  20159447                     레나      박기림        배우   
1035          재연대역  20156498                독도의 영웅들      김승헌        배우   
1038            지수  20150845    우리에게 떡볶이를 먹을 권리가 있다      박지수        배우   
1039           NaN  20161701                    위켄즈      고동민        배우   
1042            출연  20158161                   소꿉놀이      서지은        배우   
1047            장효  20141591                  한강블루스      봉만대        배우   
1048           NaN  20150258                 홀리워킹데이      곽주현        배우   
1058           NaN  20160761                    돈 맛      이은미        배우   
1070           NaN  20161961                나는 증인이다       양미        배우   
1071           NaN  20140256                    딜쿠샤      길용우        배우   
1082           NaN  20154428              야근 대신 뜨개질      김민수        배우   
1083          내레이션  20142221                     경계      문정현        배우   
1094           NaN  20140723               달에 부는 바람      김예지        배우   
1102           NaN  20158743              눈이라도 내렸으면      강성국        배우   
1130           NaN  20162970  폭력의 법칙: 나쁜 피 두 번째 이야기      김영무        배우   
1131           NaN  20166912       개인교수: 심화학습 (무삭제)      허예창        배우   
1132            승구  20166652           동거의 목적(무삭제판)      제동화        배우   
1133           NaN  20162145           청춘학당2 기생난입야사      전현수        배우   
1134            민아  20151224                     하프      안용준        배우   
1143           NaN  20162005                 맛있는 만남      차주현        배우   
1165           NaN  20153863                    그빨로       서하        배우   
1167           NaN  20156188                   블랙스톤      원태희        배우   
1168           NaN  20149496                일어서는 인간      허지원        배우   
1169           NaN  20151063                     깡치      손우혁        배우   
1177           NaN  20153803                   배짱이들      이인세        배우   
1178           NaN  20120111                 기억의 소리      정충구        배우   
1179            효주  20151181                    여교사      김하늘        배우   
1445           NaN  20160746                 파밍 보이즈      권두현        배우   
1782           NaN  20189525               젝스키스 에이틴      은지원        배우   
2265           NaN  20175587                    나비잠  나카야마 미호        배우   
2520           NaN  20184642                 어른이 되면      장혜정        배우   
2815           NaN  20196881         1919유관순-그녀들의조국      이새봄        배우   
2816           NaN  20185350                칠곡 가시나들      박금분        배우   
2817       금반지 목소리  20194045           극장판 반지의 비밀일기      방연지        배우   
2828           NaN  20190703                    우리집      김나연        배우   
2845           NaN  20198597                   교회오빠      이관희        배우   
2846           NaN  20190741                    김복동      김복동        배우   
2847            소희  20188144                 0.0MHz      정은지        배우   
2861           NaN  20166602              그대 이름은 장미      유호정        배우   
2883           NaN  20176442               자전차왕 엄복동      정지훈        배우   
2884           NaN

In [56]:
def fill_nan(i, c, x):
    leading_role.loc[i, 'peopleCd'] = c
    leading_role.loc[i, 'sex'] = x
    return

In [57]:
def fill_nan2(i, x):
    leading_role.loc[i, 'sex'] = x
    return

In [58]:
fill_nan2(506, '여자')
fill_nan2(1179, '여자')
fill_nan2(1445, '남자')
fill_nan2(1782, '남자')
fill_nan2(2265, '여자')
fill_nan2(2520, '여자')
fill_nan2(2815, '여자')
fill_nan2(2816, '여자')
fill_nan2(2817, '여자')
fill_nan2(2828, '여자')
fill_nan2(2845, '남자')
fill_nan2(2846, '여자')
fill_nan2(2847, '여자')
fill_nan2(2861, '여자')
fill_nan2(2883, '남자')
fill_nan2(2884, '여자')
fill_nan2(2931, '여자')
fill_nan2(3010, '여자') # 주연이 아닌 사람 데이터가 뽑힘. (영화코드: 20176251)

In [59]:
fill_nan(972, '20133843', '남자')
fill_nan(983, '10057553', '여자')
fill_nan(984, '10081174', '남자')
fill_nan(991, '10089613', '여자')
fill_nan(993, '10005129', '남자')
fill_nan(998, '10002465', '남자')  
fill_nan(999, '20166980', '남자') 
fill_nan(1007, '20193079', '여자') 
fill_nan(1026, '10056645', '여자') 
fill_nan(1027, '20243930', '여자')
fill_nan(1028, '20248319', '남자')
fill_nan(1029, '10026821', '남자')
fill_nan(1030, '20227826', '여자')
fill_nan(1035, '20267635', '남자')
fill_nan(1038, '20248695', '여자') 
fill_nan(1039, '20270569', '남자') 
fill_nan(1042, '20146283', '여자')  
fill_nan(1047, '10031639', '남자') 
fill_nan(1048, '20263949', '여자')
fill_nan(1058, '20069003', '여자') 
fill_nan(1070, '10089666', '여자')
fill_nan(1071, '10003464', '남자')
fill_nan(1082, '20267615', '남자') # 주연 아닌데 단역 이름만 써 있어서 뽑힘 (다큐)
fill_nan(1083, '10026800', '남자') 
fill_nan(1094, '20195668', '여자') 
fill_nan(1102, '20236706', '남자') 
fill_nan(1130, '10004967', '남자') 
fill_nan(1131, '20214447', '남자')
fill_nan(1132, '20234386', '남자') 
fill_nan(1133, '20218390', '남자') 
fill_nan(1134, '20231106', '남자')
fill_nan(1143, '20233586', '여자')
fill_nan(1165, '20214466', '여자') 
fill_nan(1167, '20113882', '남자') 
fill_nan(1168, '20234406', '남자') 
fill_nan(1169, '20227991', '남자') 
fill_nan(1177, '20214367', '남자') 
fill_nan(1178, '20253059', '남자')

In [60]:
leading_role.isnull().sum()

cast         423
movieCd        0
movieNm        0
peopleNm       0
repRoleNm      0
peopleCd       0
sex            0
dtype: int64

### leading role이 아닌 배우의 데이터가 뽑혀있는 경우, leading role 데이터로 교체
** 해당영화 movieCd: '20176251'

In [61]:
actors[actors['movieCd']=='20176251']

cast   movieCd movieNm peopleNm repRoleNm
10862    호텔직원  20176251  내안의 그놈      오윤수        배우
10863      판수  20176251  내안의 그놈      박성웅        배우
10864      동현  20176251  내안의 그놈       진영        배우
10865      미선  20176251  내안의 그놈      라미란        배우
10866      현정  20176251  내안의 그놈      이수민        배우
10867      만철  20176251  내안의 그놈      이준혁        배우
10868      종기  20176251  내안의 그놈      김광규        배우
10869      서연  20176251  내안의 그놈      민지아        배우
10870     양사장  20176251  내안의 그놈      윤경호        배우
10871      재익  20176251  내안의 그놈      김현목        배우
10872      철호  20176251  내안의 그놈      권혁범        배우
10873      태욱  20176251  내안의 그놈      이호수        배우
10874  판수조직원2  20176251  내안의 그놈      김상겸        배우
10875    문학교사  20176251  내안의 그놈      조성희        배우
10876     재희   20176251  내안의 그놈      박경혜        배우
10877    국사교사  20176251  내안의 그놈      장진희        배우
10878   민우친구8  20176251  내안의 그놈      박지훈        배우
10879    일어교사  20176251  내안의 그놈      정윤하        배우
10880   민우친구7  20176251  내안의 그놈      정원창        배우

데이터의 배열이 잘못 되어 있어 단역배우 '오윤수'가 주연배우 '박성웅' 윗줄에 배치되었던 것.  
leading_role의 오윤수 데이터를 박성웅 데이터로 교체한다.

In [62]:
leading_role.loc[3010]

cast             호텔직원
movieCd      20176251
movieNm        내안의 그놈
peopleNm          오윤수
repRoleNm          배우
peopleCd     20329332
sex                여자
Name: 3010, dtype: object

In [63]:
leading_role.loc[3010, 'cast'] = '판수'
leading_role.loc[3010, 'peopleNm'] = '박성웅'
leading_role.loc[3010, 'peopleCd'] = '10028773'
leading_role.loc[3010, 'sex'] = '남자'

In [64]:
leading_role.loc[3010]

cast               판수
movieCd      20176251
movieNm        내안의 그놈
peopleNm          박성웅
repRoleNm          배우
peopleCd     10028773
sex                남자
Name: 3010, dtype: object

In [68]:
leading_role.sample(3)

cast   movieCd                   movieNm peopleNm repRoleNm  \
1221       수완  20157631                  다른 길이 있다      김재욱        배우   
2793  이리샤 목소리  20181096                 마왕의 딸 이리샤      천우희        배우   
1671      NaN  20174643  국정교과서 516일 : 끝나지 않은 역사전쟁      한홍구        배우   

      peopleCd sex  
1221  10005508  남자  
2793  20133966  여자  
1671  20284549  남자

In [70]:
leading_role.to_csv('../01.Data_Collection/actor_leading_role.csv', index=False)